misago/legal/models.py

```diff
@@ -1 +1,113 @@
+
+
+class NoActiveLegalAgreement(ValidationError):
+    pass
+
+
+class LegalAgreementManager(models.Manager):
+    def get_current(self):
+        try:
+            return self.filter(is_active=True)
+        except self.model.DoesNotExist:
+            raise NoActiveLegalAgreement(
+                u'Please create an active LegalAgreement'
+            )
+
+
+class LegalAgreement(models.Model):
+    created_at = models.DateTimeField(default=timezone.now)
+    modified_at = models.DateTimeField(auto_now=True)
+    published_at = models.DateTimeField(blank=True, null=True)
+    key = models.SlugField(unique=True, blank=True)
+
+    TOS = 'terms_of_service'
+    PRIVACY = 'privacy_policy'
+    LEGAL_CHOICES = (
+        (TOS, 'Terms of service'),
+        (PRIVACY, 'Privacy Policy'),
+    )
+    title = models.CharField(
+        max_length=20,
+        choices=LEGAL_CHOICES,
+        default=TOS
+    )
+    text = models.TextField(
+        verbose_name=_('content'),
+        blank=True
+    )
+    is_active = models.BooleanField(
+        default=False,
+        verbose_name=_('active/publish'),
+        help_text=_(
+            u'Only one terms of service is allowed to be active'
+        )
+    )
+
+    objects = LegalAgreementManager()
+
+    def __init__(self, *args, **kwargs):
+        super(LegalAgreement, self).__init__(*args, **kwargs)
+        self.old_is_active = self.is_active
+
+    def __str__(self):
+        return self.title + "-" + self.created_at.strftime("%Y-%m-%d")
+
+    def has_user_agreed_latest_legal(self, user, legal):
+        return UserAgreementAcceptance.objects.filter(
+            legal_agreement=LegalAgreement.objects.get_current().get(title=legal),
+            user=user,
+        ).exists()
+
+    def accept(self, user):
+        acceptance = UserAgreementAcceptance(
+            legal_agreement=LegalAgreement.objects.get_current().get(title=self.title), user=user)
+        acceptance.save()
+
+    def save(self, *args, **kwargs):
+        """ Ensure that one LegalAgreement is active"""
+        if  self.is_active:
+            LegalAgreement.objects.exclude(id=self.id).filter(title=self.title).update(is_active=False)
+        else:
+            if not LegalAgreement.objects.exclude(id=self.id).filter(is_active=True, title=self.title).exists():
+                raise NoActiveLegalAgreement(
+                    u'One of the {} must be marked active'.format(self.title)
+                )
+
+        # Set the latest published date.
+        if self.old_is_active != self.is_active and self.is_active:
+            self.published_at = timezone.now()
+
+        self.key = slugify(self.title + " " + self.created_at.strftime("%Y-%m-%d"))
+        super(LegalAgreement, self).save(*args, **kwargs)
+
+
+class UserAgreementAcceptance(models.Model):
+    accepted_on = models.DateTimeField(auto_now=True)
+    has_accepted = models.BooleanField(default=False)
+    user_ip = models.GenericIPAddressField()
```

Comment:
Lets not record IP used to give this permission ;) I know recording IP everywhere was running trend in Misago, but I've recently decided to limit number of places this is being done to avoid possible headaches with GDPR audits.